### ISSUE TITLE
cargo occasionally fails to download any packages due to "no hash listed for"

### ISSUE THREAD

**Issue by jmesmon** [rust-lang/cargo#1355]
When `cargo build` is re-run, various packages fail in the same way (typically a different one each time).

An example:
```
cody@two /home/cody/foo % cargo -V         
cargo 0.0.1-pre-nightly (43755c0 2015-02-19) (built 2015-02-21)
cody@two /home/cody/foo % cargo build
    Updating registry `https://github.com/rust-lang/crates.io-index`
 Downloading env_logger v0.2.2
Unable to get packages from source

To learn more, run the command again with --verbose.
cody@two /home/cody/foo % cargo build --verbose
    Updating registry `https://github.com/rust-lang/crates.io-index`
 Downloading docopt v0.6.39
Unable to get packages from source

Caused by:
  Failed to download package `docopt v0.6.39` from https://crates.io/api/v1/crates/docopt/0.6.39/download

Caused by:
  no hash listed for docopt v0.6.39
cody@two /home/cody/foo % 
```

**Comment by jmesmon** [comment]
Persists even when `$HOME/.cargo` is renamed

**Comment by jmesmon** [comment]
Renaming my repo local `.cargo/config`, starting a build (that downloaded some, but not all packages), and then killing the build and moving my `.cargo/config` back into place & retrying the build worked around this issue.

Contents of my repo local `.cargo/config`:

```
#paths = [ "../../g/rust-openssl" ]
paths = [ "../rust-dev", "../../g/rustc-serialize", "../../g/rust-phf", "../../g/r2d2", "../../g/rust-openssl", "../../g/rust-postgres", "../../g/xdg-rs", "../../g/tempdir" ]
```

**Comment by sfackler** [comment]
I'm running into this as well.

**Comment by jmesmon** [comment]
This might be completely off track, but: In my case, some of the paths in the .cargo/config of the package I was building also had .cargo/config's in them, which also had some paths set.

**Comment by sfackler** [comment]
I only have one config in the directory all of my cargo packages live, so I don't think it's that.

**Comment by Byron** [comment]
I am now seeing the same issue:

```bash
➜  yup-oauth2 git:(next) ✗ cargo test --verbose
    Updating registry `https://github.com/rust-lang/crates.io-index`
 Downloading open v1.0.1
Unable to get packages from source

Caused by:
  Failed to download package `open v1.0.1` from https://crates.io/api/v1/crates/open/1.0.1/download

Caused by:
  no hash listed for open v1.0.1
```

It's new to me, and just occurred after publishing `open` v1.0.1. The latter worked without any issues whatsoever.

Reproducing it should be easy by just adding `open = "*"` as project dependency.
For now I could workaround this issue by adding yet another path-override.

**Comment by zr40** [comment]
I'm also seeing this with `router` 0.0.9.

https://crates.io/api/v1/crates/router/0.0.9/download returns an HTML error page with the message "Oops, that route doesn't exist!".

**Comment by alexcrichton** [comment]
Are there some example crates I could test out building to reproduce this? I haven't seen this yet and have been unable to reproduce locally.

**Comment by Byron** [comment]
Unfortunately, I also can't reproduce this anymore, even when setting the version of open to *1.0.1*, which triggered the error in my example.
I will let you know if I encounter it again.

**Comment by zr40** [comment]
I also haven't seen this error recently. Like an earlier comment mentioned, I too had a number of overrides in `~/.cargo/config` at the time. But restoring these doesn't appear to be sufficient to reproduce this error.

**Comment by tomaka** [comment]
I've encountered this situation as well.

Here is the situation:
 - My binary depends on the vorbis and cpal crates
 - cpal has a dev-dependency on the vorbis crate

I added a local override for `cpal` in my binary, ran `cargo update` followed by `cargo build`, and got the error with `vorbis 0.0.9`.

**Comment by zr40** [comment]
I just hit this bug again. I did the following after leaving my project untouched for a few days:

1. `rustup.sh`
2. `cargo update`
3. `cargo build`

This failed with the following output:
```
[..]
 Downloading phf_macros v0.7.0
Unable to get packages from source

Caused by:
  Failed to download package `phf_macros v0.7.0` from https://crates.io/api/v1/crates/phf_macros/0.7.0/download

Caused by:
  no hash listed for phf_macros v0.7.0
```

I have the following in `.cargo/config`:
```
paths = ["/Users/zr40/git/rust-postgres-array"]
```
This repository is at sfackler/rust-postgres-array@3216821.

When I removed `.cargo/config`, `cargo build` succeeded. Restoring `.cargo/config` is not sufficient to reproduce, I also needed to delete `.cargo/registry/cache/github.com-1ecc6299db9ec823/phf_macros-0.7.0.crate` to make `cargo build` fail in downloading that crate again.

This is the `Cargo.toml` and `Cargo.lock` of this project: https://gist.github.com/zr40/c422efe6c197db2b1b85
